cachelib/navy/common/FdpNvme.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cachelib/navy/common/FdpNvme.h"
+
+#include <linux/nvme_ioctl.h>
+#include <sys/ioctl.h>
+#include <errno.h>
+#include <cstring>
+#include <numeric>
+
+namespace facebook {
+namespace cachelib {
+namespace navy {
+
+FdpNvme::FdpNvme(int fd) {
+  nvmeData_ = readNvmeInfo(fd);
+  initializeFDP(fd);
+  XLOG(INFO)<< "Creating FdpDevice on fd :" << fd;
+}
+
+int FdpNvme::allocateFdpHandle() {
+  uint16_t phndl;
+
+  // Get NS specific Fdp Placement Handle(PHNDL)
+  if (nextPIDIdx_ <= maxPIDIdx_) {
+    phndl = nextPIDIdx_++;
+  } else {
+    phndl = kDefaultPIDIdx;
+  }
+  // Get Device specific Fdp Placement ID for PHNDL
+  auto pid = getFdpPID(phndl);
+
+  return static_cast<int>(pid);
+}
+
+void FdpNvme::initializeFDP(int fd) {
+  nextPIDIdx_ = kDefaultPIDIdx + 1;
+
+  Buffer buffer = nvmeFdpStatus(fd);
+  if (!buffer.isNull()) {
+    struct nvme_fdp_ruh_status *ruh_status =
+      reinterpret_cast<struct nvme_fdp_ruh_status*>(buffer.data());
+
+    if (ruh_status->nruhsd) {
+      maxPIDIdx_ = ruh_status->nruhsd - 1;
+      for (uint16_t i = 0; i <= maxPIDIdx_; ++i) {
+        placementIDs_[i] = ruh_status->ruhss[i].pid;
+      }
+    }
+  }
+  XLOG(INFO)<< "Creating NvmeFdp, fd :"<< fd<<" Num of PID: "
+            <<maxPIDIdx_ + 1<<" First PID: "<<placementIDs_[0]
+            <<" Last PID: "<<placementIDs_[maxPIDIdx_];
```

Comment:
This looks like a debug info. `XLOG(DEBUG)` or so. Also, can you use formatted string form? i.e., `XLOGF(DEBUG, ...)`?